### PR TITLE
New unit tests for TiffHeader and Cr2Header

### PR DESCRIPTION
**Pull request by piponazo**
In this PR I am adding few characterisation tests for the **TiffHeader** and **Cr2Header** classes. I am doing this work while studying the code, since I think that it is a good way to add:

- Characterisation tests for defining the current behaviour of the code.
- Probably these can be considered as formal unit tests for those classes, since I do not think we change the design of these classes in the future.
- Add documentation for the code.

While adding these tests I found some minor issues in the library code that I had to fix to make the tests pass. I also removed some old stuff from the .gitignore file and added some other things that are annoying me during my normal development workflow.

I think that this code will not work on Windows. These are internal classes of the exiv2lib and we are not exporting the symbols (what it is correctly done). In order to be able to merge this PR I will need to take care of the issue described in #149. I see a potential risks in addressing that issue:

- If I use the CMake **OBJECT** targets for reusing the object files for the targets **exiv2lib** and **unit_tests**, maybe the CMake-XCode generator would not be able to handle that (we could open an issue for fixing that in the future).

### CODE REVIEW

File: src/cr2header_int.hpp [hunk 1]
```diff
@@ -19,14 +19,13 @@
...
-  @file    cr2image_int.hpp
+  @file    cr2header_int.hpp
   @brief   Internal classes to support CR2 image format
   @author  Andreas Huggel (ahu)
            <a href="mailto:ahuggel@gmx.net">ahuggel@gmx.net</a>
   @date    23-Apr-08, ahu: created
  */
-#ifndef CR2IMAGE_INT_HPP_
-#define CR2IMAGE_INT_HPP_
+#pragma once
```
Comment (D4N) [inline src/cr2header_int.hpp]: Please no `#pragma once`, the include guards do their job just fine and are the standard way of ensuring only one inclusion.

File: src/cr2image.cpp [hunk 2]
```diff
@@ -247,7 +247,7 @@ namespace Exiv2 {
 
     bool Cr2Header::read(const byte* pData, uint32_t size)
     {
-        if (size != 16)
+        if (!pData || size != 16)
```
Comment (D4N) [inline src/cr2image.cpp]: Good catch, we don't want to dereference null pointers.

File: unitTests/CMakeLists.txt [hunk 3]
```diff
@@ -8,10 +8,11 @@ add_executable(unit_tests mainTestRunner.cpp
 target_link_libraries(unit_tests
     PRIVATE
         exiv2lib
+        $<TARGET_OBJECTS:exiv2lib_int>
```
Comment (D4N) [inline unitTests/CMakeLists.txt]: Travis seems to choke on this addition.

**Review by D4N — CHANGES_REQUESTED** [review]
I don't like the `#pragma once` and that the CI is broken, beside that I have no objections.

File: unitTests/test_cr2header.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+TEST(ACr2Header, failToReadDataFromBufferWithSizeDifferentThan16)
+{
+    Internal::Cr2Header header;
+    ASSERT_FALSE(header.read(NULL, 15));
+    ASSERT_FALSE(header.read(NULL, 17));
+}
+
+TEST(ACr2Header, failToReadDataFromBufferWithInvalidByteOrder)
+{
+    static const byte bufferInvalidByteOrder[] =  {0x88, 0x49, 0x2a, 0x00, 0x10, 0x00, 0x00, 0x00,
```
Comment (D4N) [inline unitTests/test_cr2header.cpp]: Is there a particular reason why you are declaring these buffers as static? (I am asking out of curiosity, is this required by gtest?)

**Comment by D4N** [comment]
Very nice job @piponazo, thank you for increasing the test coverage! I can confirm that the code builds and the tests pass on Fedora 26. So I think travis has some problems due to an old cmake version.

File: unitTests/test_cr2header.cpp (hunk 4 above)
Comment (piponazo) [reply unitTests/test_cr2header.cpp]: ups, no, it is not needed. I was lazy and I copied/pasted from line 6. I will remove the static from here.

**Comment by piponazo** [comment]
Thanks @D4N. I will not work on this PR until I solve the situation detected in #178. I actually created the other PR to first solve the issue about testing private parts of the libraries, and then we can only focus on the tests and changes in the library code.

I'm going to decline this PR and I will reopen later.

**Pull request closed without merging**